### PR TITLE
CLI: fixup shutdown/interrupt code

### PR DESCRIPTION
**Pull request by stefanpenner**
- [x] fix tests
- [x] add regression test

### CODE REVIEW

File: tests/unit/cli/cli-test.js [hunk 1]
```diff
@@ -228,6 +235,31 @@ describe('Unit: CLI', function() {
     });
   });
 
+  describe.only('command crash', function() {
```
Comment (chadhietala) [inline tests/unit/cli/cli-test.js]: Remove the `only`

Comment (stefanpenner) [inline tests/unit/cli/cli-test.js]: Ooo. Will fix

**Comment by stefanpenner** [comment]
Windows CI seems unhappy, re-running. If not I can take a look tonight when i have a windows machine handy.

File: tests/unit/cli/cli-test.js [hunk 2]
```diff
@@ -228,6 +235,31 @@ describe('Unit: CLI', function() {
...
+      });
+
+      project.eachAddonCommand = function(callback) {
+        callback('custom-addon', {
+          custom: CustomCommand,
+        });
+      };
+
+      return ember(['custom']).then(function() {
+        expect(false, 'should have rejected').to.eql(false);
```
Comment (ro0gr) [inline tests/unit/cli/cli-test.js]: Shouldn't one of these false be a true? Seems like this line will always pass if a promise isn't rejected.

Comment (stefanpenner) [inline tests/unit/cli/cli-test.js]: 👍

Comment (Turbo87) [inline tests/unit/cli/cli-test.js]: actually we should use https://github.com/domenic/chai-as-promised for tests like that. e.g.

```js
return expect(ember(['custom'])).to.be.rejected;
```

File: tests/unit/cli/cli-test.js [hunk 3]
```diff
@@ -228,6 +235,31 @@ describe('Unit: CLI', function() {
...
+      });
+
+      project.eachAddonCommand = function(callback) {
+        callback('custom-addon', {
+          custom: CustomCommand,
+        });
+      };
+
+      return ember(['custom']).then(function() {
+        expect(false, 'should have rejected').to.eql(true);
```
Comment (stefanpenner) [inline tests/unit/cli/cli-test.js]: switch to chai-as-promised.

**Comment by ro0gr** [comment]
I've tested this pr against several known issues:
 - https://github.com/ember-fastboot/ember-cli-fastboot/issues/386
 - https://github.com/emberjs/data/pull/4965

The good news is that these issues are addressed.

The bad news is that it breaks testem integration on windows.

I've noticed it in ["ember new foo, test, SIGINT exits with error and clears tmp/"](https://github.com/ember-cli/ember-cli/blob/93c4fd5763fa69fcb11eb0bee1b6b38e57ad4d15/tests/acceptance/smoke-test-slow.js#L302) test failure. The test on itself passes but [afterEach](https://github.com/ember-cli/ember-cli/blob/93c4fd5763fa69fcb11eb0bee1b6b38e57ad4d15/tests/acceptance/smoke-test-slow.js#L43) hook fails with the following message:

>  √ ember new foo, test, SIGINT exits with error and clears tmp/ (145694ms)
>   1) "after each" hook for "ember new foo, test, SIGINT exits with error and clears tmp/"
>
>  1 passing (4m)
 > 1 failing
 >
 > 1) Acceptance: smoke-test "after each" hook for "ember new foo, test, SIGINT exits with error and clears tmp/":
     AssertionError: expected "C:\Users\kudgo\workspace\ember-cli\tmp\some_cool_app_clone-KeX18ufj.tmp" to not exist

Literally `afterEach` is unable to delete temporary application made for test. The reason for this is that testem doesn't kill `phantomjs` process anymore. This process locks test application directory.
 
<details>
<summary>testem.log - before this pr</summary>

> info Test'em starting..
info Seeking for config file...
info Starting ci
info Starting server
info Waiting for tests.
info Start triggered test run.
info Running tests...
info Tests running.
info spawning: phantomjs - [ 'C:\\Users\\kudgo\\workspace\\ember-cli\\node_modules\\testem\\assets\\phantom.js',
  'http://localhost:7357/5383/tests/index.html?hidepassed' ]
info New client connected: PhantomJS 2.1 5383
info tryAttach PhantomJS 2.1 5383
info received SIGINT signal
info Closing browser PhantomJS.
WARN PhantomJS closed 1
info Process PhantomJS terminated. 1 null
</details>

<details>
<summary>testem.log - this pr</summary>

> info Test'em starting..
info Seeking for config file...
info Starting ci
info Starting server
info Waiting for tests.
info Start triggered test run.
info Running tests...
info Tests running.
info spawning: phantomjs - [ 'C:\\Users\\kudgo\\workspace\\ember-cli\\node_modules\\testem\\assets\\phantom.js',
  'http://localhost:7357/5096/tests/index.html?hidepassed' ]
info New client connected: PhantomJS 2.1 5096
info tryAttach PhantomJS 2.1 5096
info
</details>



In the second log we miss `received SIGINT signal` entry which causes the problem.

I'm in process of investigation. Any ideas would be awesome.

**Comment by ro0gr** [comment]
It's not just a test setup issue. It also reproducible by manual verification

**Comment by ro0gr** [comment]
> The bad news is that it breaks testem integration on windows.

After some testing it turned out that it breaks only with `phantomjs`. `Chrome` launcher is disposed properly.
Seems like some kind of timing issue. Wrapping of interruption handler in timeout "fixes" the problem:
```js
        removeInterruptionHandler = onProcessInterrupt.addHandler(() => {
          return new RSVP.Promise(function(resolve) {
            setTimeout(() => {
              command.onInterrupt();
              resolve();
            }, 1000);
          });
```
But it seems like a bad solution.

I've checked how does "ctrl+c" work against `ember-cli@2.12` the problem persist there. The reason why this issue didn't appear earlier in tests is that we didn't have smoke test for `ember test` interruption.

This pr looks good for me but if we merge most likely it will break windows CI.

File: lib/cli/cli.js [hunk 4]
```diff
@@ -199,19 +193,10 @@ class CLI {
...
-      return runPromise;
-    })
-    .finally(() => {
-      if (shutdownOnExit) {
-        shutdownOnExit();
-      }
-    })
-    .catch(this.logError.bind(this));
+    }).finally(() => {
+      onProcessInterrupt.teardown();
```
Comment (ro0gr) [inline lib/cli/cli.js]: What is the reason for additional `teardown` here? https://github.com/ember-cli/ember-cli/pull/7019/files#diff-16a41bfbe414387d773df099f941217eR162 should do all the necessary cleanup?

**Comment by ro0gr** [comment]
@stefanpenner I believe now with headless chrome this should pass under appveyor

**Comment by kellyselden** [comment]
@ro0gr Rebased on master.

**Comment by ro0gr** [comment]
@kellyselden great!

There is one failing test which breaks [here ](https://github.com/ember-cli/ember-cli/blob/7274291ae515c53a5de6924dfe27797568e7b20c/tests/unit/utilities/will-interrupt-process-test.js#L191).

I think it should be re-written to match updated `will-interrupt-process` api. 

**Before**:
```js
      // ...
      removewillInterruptProcess.addHandler(cb);
      td.verify(process.stdin.setRawMode(true));

      willInterruptProcess.removeHandler(cb);
      td.verify(process.stdin.setRawMode(false));

      td.verify(process.stdin.setRawMode(), {
        ignoreExtraArgs: true,
        times: 2,
      });
```

**After**:
```js
      // ...
      let removeHandler = willInterruptProcess.addHandler(cb); // `addHandler` now yields `removeHandler`
      td.verify(process.stdin.setRawMode(true));

      removeHandler(cb);
      td.verify(process.stdin.setRawMode(false));

      td.verify(process.stdin.setRawMode(), {
        ignoreExtraArgs: true,
        times: 2,
      });
```
I believe this should work.

-----
Saying the truth I find this test fragile. We could test it much more better if merge https://github.com/ember-cli/ember-cli/pull/7086 first.

**Comment by stefanpenner** [comment]
@ro0gr its merged!

**Comment by Turbo87** [comment]
closing due to inactivity and conflicts

**Pull request closed without merging**